src/main/java/org/candlepin/pki/impl/BouncyCastlePKIUtility.java

```diff
@@ -0,0 +1,650 @@
+     *
+     * @param algorithm
+     *  the algorithm used to generate the key
+     *
+     * @return
+     *  a PublicKey instance
+     */
+    private PublicKey generatePublicKey(byte[] keydata, String algorithm)
+        throws NoSuchAlgorithmException, InvalidKeySpecException {
+
+        KeyFactory factory = KeyFactory.getInstance(algorithm, BouncyCastleProviderLoader.getProvider());
+        X509EncodedKeySpec spec = new X509EncodedKeySpec(keydata, algorithm);
+
+        return factory.generatePublic(spec);
+    }
+
+    /**
+     * Generates a PrivateKey instance from the provided key data and algorithm.
+     *
+     * @param keydata
+     *  the PKCS8 private key data from which to generate a PrivateKey instance
+     *
+     * @param algorithm
+     *  the algorithm used to generate the key
+     *
+     * @return
+     *  a PrivateKey instance
+     */
+    private PrivateKey generatePrivateKey(byte[] keydata, String algorithm)
+        throws NoSuchAlgorithmException, InvalidKeySpecException {
+
+        KeyFactory factory = KeyFactory.getInstance(algorithm, BouncyCastleProviderLoader.getProvider());
+        PKCS8EncodedKeySpec spec = new PKCS8EncodedKeySpec(keydata, algorithm);
+
+        return factory.generatePrivate(spec);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public KeyPair generateKeyPair() throws KeyException {
+        try {
+            KeyPairGenerator keyGen = KeyPairGenerator.getInstance(KEY_ALGORITHM);
+            keyGen.initialize(KEY_SIZE);
+            KeyPair keypair = keyGen.generateKeyPair();
+            return this.buildInsecureKeyPair(keypair);
+        }
+        catch (NoSuchAlgorithmException e) {
+            throw new KeyException(e);
+        }
+    }
+
+    /**
+     * Rebuilds an "insecure" KeyPair instance from another KeyPair instance. The returned KeyPair
+     * will allow fetching the encoded form of both the public and private key.
+     *
+     * @param keypair
+     *  the keypair to rebuild
+     *
+     * @return
+     *  a keypair which allows extracting the encoded form of both the public and private key
+     */
+    private KeyPair buildInsecureKeyPair(KeyPair keypair) throws KeyException {
+        return this.buildInsecureKeyPair(keypair.getPublic(), keypair.getPrivate());
+    }
+
+    /**
+     * Builds an "insecure" KeyPair instance from the provided public and private keys. The returned
+     * KeyPair will allow fetching the encoded form of both the public and private key.
+     *
+     * @param publicKey
+     *  the public key to use to build the insecure keypair
+     *
+     * @param privateKey
+     *  the private key to use to build the insecure keypair
+     *
+     * @return
+     *  a keypair which allows extracting the encoded form of both the public and private key
+     */
+    private KeyPair buildInsecureKeyPair(PublicKey publicKey, PrivateKey privateKey)
```

Comment:
Do we need the concept of "insecure" keypairs with BouncyCastle? This path is likely obsolete with BC.

The wrapping and unwrapping logic came about because JSS's key implementations disallow getting the raw key bytes at runtime, so we needed to do this to get them. If BC doesn't have the same restriction, we need not bother with the slow wrap/unwrap cycle to get it.